core/src/replay_stage.rs

```diff
@@ -1285,8 +1285,14 @@ impl ReplayStage {
     ) -> (ProgressMap, HeaviestSubtreeForkChoice) {
         let (root_bank, frozen_banks, duplicate_slot_hashes) = {
             let bank_forks = bank_forks.read().unwrap();
+            let root_bank = bank_forks.root_bank();
             let duplicate_slots = blockstore
-                .duplicate_slots_iterator(bank_forks.root_bank().slot())
+                // It is important that the root bank is not marked as duplicate on initialization.
+                // Although this bank could contain a duplicate proof, the fact that it was rooted
+                // either during a previous run or artificially means that we should ignore any
+                // duplicate proofs for the root slot, thus we start consuming duplicate proofs
+                // from the root slot + 1
+                .duplicate_slots_iterator(root_bank.slot().saturating_add(1))
```

Comment:
This is the only functional change, remaining is test code.

This backport ensures that nodes are able to restart cleanly in the presence of duplicate blocks. The root chosen upon restart could have previously been a duplicate block that was resolved, or chosen as a restart slot anyway. It is important that we ignore the duplicate column for the restart root slot, otherwise we will panic on replay initialization.